src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -50,6 +51,176 @@ static void Main()
                 );
         }
 
+        [Fact]
+        public void SpeakableInference_MethodTypeInference()
+        {
+            var source =
+@"class Program
+{
+    void M<T>(T t)
+    {
+        if (t == null) throw null;
+        t.ToString();
+        var t2 = Copy(t);
+        t2.ToString(); // warn
+    }
+    static T Copy<T>(T t) => throw null;
+}";
+            var comp = CreateCompilationWithIndexAndRange(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (8,9): warning CS8602: Possible dereference of a null reference.
+                //         t2.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t2").WithLocation(8, 9)
+                );
+        }
+
+        [Theory]
+        [InlineData("void M(int? t)")]
+        [InlineData("void M<T>(T? t) where T : struct")]
+        public void SpeakableInference_MethodTypeInference_NullableValueType(string signature)
+        {
+            var source =
+@"class Program
+{
+    SIGNATURE
+    {
+        if (t == null) throw null;
+        t.Value.ToString();
+        var t2 = Copy(t);
+        t2.Value.ToString(); // warn
+    }
+    static T Copy<T>(T t) => throw null;
+}";
+            var comp = CreateCompilationWithIndexAndRange(source.Replace("SIGNATURE", signature), options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (8,9): warning CS8629: Nullable value type may be null.
+                //         t2.Value.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "t2.Value").WithLocation(8, 9)
+                );
+        }
+
+        [Fact]
+        public void SpeakableInference_ArrayTypeInference()
+        {
+            var source =
+@"class Program
+{
+    void M<T>(T t)
+    {
+        if (t == null) throw null;
+        t.ToString();
+        var t2 = new[] { t };
+        t2[0].ToString(); // warn
+    }
+}";
+            var comp = CreateCompilationWithIndexAndRange(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (8,9): warning CS8602: Possible dereference of a null reference.
+                //         t2[0].ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t2[0]").WithLocation(8, 9)
+                );
+        }
+
+        [Theory]
+        [InlineData("void M(int? t)")]
+        [InlineData("void M<T>(T? t) where T : struct")]
+        public void SpeakableInference_ArrayTypeInference_NullableValueType(string signature)
+        {
+            var source =
+@"class Program
+{
+    SIGNATURE
```

Comment:
>SIGNATURE [](start = 4, length = 9)

Consider writing two tests.

Reply:
My 2 cents: I like how theories can keep our test assets more compact.

---
In reply to: [242669373](https://github.com/dotnet/roslyn/pull/31813#discussion_r242669373) [](ancestors = 242669373)

Reply:
Some context: I expected there would be more than two variants here ;-)